File: typescript/src/ships/mobility.ts
/**
 * Data-free speed and handling calculations for a loaded ship. Speed and angular-rate
 * endpoints interpolate linearly by the ENG-PIP fraction before thruster mass curves
 * are applied.
 *
 * Ported from EDCD/Coriolis and cross-checked against EDSY's mass-curve calculation;
 * see [`ATTRIBUTIONS.md`](https://github.com/DarkSession/Elite-Dangerous-Almanac/blob/main/ATTRIBUTIONS.md).
 *
 * @packageDocumentation
 */

import {
    massCurveMultiplier,
    requireFiniteNonNegative,
    validateMassCurve,
} from './internal/mass-curve.js';

/**
 * One post-engineering thruster performance curve.
 *
 * Masses must be strictly ordered `minMass < optMass < maxMass`. The all-equal
 * mass form is accepted only for a constant curve. Multipliers must be strictly
 * ordered `minMultiplier < optMultiplier < maxMultiplier`, or all equal for a
 * constant curve.
 */
export interface ThrusterCurveParams {
    /** Finite non-negative loaded mass at which performance reaches {@link maxMultiplier}, in tonnes. */
    readonly minMass: number;
    /** Finite non-negative loaded mass at which performance is exactly {@link optMultiplier}, in tonnes. */
    readonly optMass: number;
    /** Finite non-negative maximum loaded mass the thrusters can move, in tonnes. */
    readonly maxMass: number;
    /** Finite non-negative performance multiplier at {@link maxMass}. */
    readonly minMultiplier: number;
    /** Finite non-negative performance multiplier at {@link optMass}. */
    readonly optMultiplier: number;
    /** Finite non-negative performance multiplier at {@link minMass}. */
    readonly maxMultiplier: number;
}

/** A fitted thruster's post-engineering mass curves. */
export interface ThrusterParams extends ThrusterCurveParams {
    /** Optional speed/boost curve; enhanced-performance thrusters differ from handling. */
    readonly speedCurve?: ThrusterCurveParams;
    /** Optional pitch/roll/yaw curve; enhanced-performance thrusters differ from speed. */
    readonly rotationCurve?: ThrusterCurveParams;
}

/** Everything {@link mobilityMetrics} needs about one loaded ship. */
export interface MobilityInput {
    /** Finite non-negative hull speed at multiplier `1` and zero ENG pips, in m/s. */
    readonly minimumSpeed: number;
    /** Finite hull speed at multiplier `1` and four ENG pips, in m/s, at least {@link minimumSpeed}. */
    readonly maximumSpeed: number;
    /** Finite non-negative hull boost speed at the thruster curve's `1` multiplier, in m/s. */
    readonly boost: number;
    /** Finite non-negative pitch rate at multiplier `1` and four ENG pips, in °/s. */
    readonly pitch: number;
    /** Finite hull pitch rate at zero ENG pips, in °/s from `0` through {@link pitch}. */
    readonly minPitch: number;
    /** Finite non-negative roll rate at multiplier `1` and four ENG pips, in °/s. */
    readonly roll: number;
    /** Finite hull roll rate at zero ENG pips, in °/s from `0` through {@link roll}. */
    readonly minRoll: number;
    /** Finite non-negative yaw rate at multiplier `1` and four ENG pips, in °/s. */
    readonly yaw: number;
    /** Finite hull yaw rate at zero ENG pips, in °/s from `0` through {@link yaw}. */
    readonly minYaw: number;
    /** Finite non-negative loaded mass — hull, modules, fuel and cargo — in tonnes. */
    readonly mass: number;
    /** The fitted thrusters' post-engineering curve, or no curve when none are fitted. */
    readonly thrusters?: ThrusterParams | null;
    /** Finite pips assigned to ENG, in `[0, 4]`. Defaults to `4`. */
    readonly enginesPips?: number;
}

/**
 * Speed and rotation rates for one loaded ship.
 *
 * @remarks
 * Frozen — nested records and lists included — so a result can be held, cached and
 * shared without a defensive copy. Derive a changed figure with a spread rather than
 * by assigning into one.
 */
export interface MobilityMetrics {
    /**
     * The loaded mass these figures were calculated at, in tonnes — the
     * {@link MobilityInput.mass} that went in.
     *
     * @remarks
     * Reported so a build-level caller, which never states the mass itself, can read
     * what the curve was evaluated at. Against the fitted curve's `optMass` and
     * `maxMass` it is the build's position on that curve: at or below `optMass` the
     * thrusters are at or above their rated performance, and past `maxMass` they do not
     * move the ship at all — {@link massCurveMultiplier} is `0` there.
     */
    readonly loadedMass: number;
    /** Top speed at this mass and ENG allocation, in metres per second. */
    readonly speed: number;
    /** Boost speed at this mass, in metres per second. */
    readonly boost: number;
    /** Pitch rate at this mass and ENG allocation, in degrees per second. */
    readonly pitch: number;
    /** Roll rate at this mass and ENG allocation, in degrees per second. */
    readonly roll: number;
    /** Yaw rate at this mass and ENG allocation, in degrees per second. */
    readonly yaw: number;
    /** The thruster mass curve's performance multiplier at this loaded mass. */
    readonly massCurveMultiplier: number;
    /** The rotation curve's multiplier; differs for enhanced-performance thrusters. */
    readonly rotationMassCurveMultiplier: number;
}

const requireFiniteRange = (
    scope: string,
    name: string,
    value: number,
    minimum: number,
    maximum: number,
): void => {
    if (!Number.isFinite(value) || value < minimum || value > maximum) {
        throw new RangeError(
            `${scope}: ${name} must be a finite number from ${minimum} to ${maximum}`,
        );
    }
};

const curveMultiplier = (scope: string, mass: number, thrusters: ThrusterCurveParams): number =>
    massCurveMultiplier({ scope, mass: 'mass', curve: 'thrusters' }, mass, thrusters);

/**
 * Resolve a thruster's performance multiplier at a loaded mass.
 *
 * @param mass - Loaded mass in tonnes.
 * @param thrusters - The fitted thruster curve, post-engineering.
 * @returns The curve multiplier, or `0` above the thrusters' maximum supported mass.
 * @throws {RangeError} If mass or a curve value is not finite and non-negative, or
 * the curve masses or multipliers do not follow the ordering documented by
 * {@link ThrusterCurveParams}.
 * @example
 * ```ts
 * import { thrusterMassCurveMultiplier } from '@elite-dangerous-almanac/core/ships/mobility';
 *
 * thrusterMassCurveMultiplier(48, {
 *   minMass: 24, optMass: 48, maxMass: 72,
 *   minMultiplier: 0.83, optMultiplier: 1, maxMultiplier: 1.03,
 * }); // -> 1
 * ```
 */
export function thrusterMassCurveMultiplier(mass: number, thrusters: ThrusterCurveParams): number {
    return curveMultiplier('thrusterMassCurveMultiplier', mass, thrusters);
}

/**
 * Calculate a loaded ship's top speed, boost speed and rotation rates.
 *
 * @remarks
 * Speed, pitch, roll and yaw interpolate linearly from their installed zero-ENG-PIP
 * endpoints to their four-PIP endpoints. The fitted thruster's speed and rotation
 * mass-curve multipliers are then applied to those interpolated hull values. Boost is
 * independent of ENG allocation and uses the speed curve at the loaded mass.
 *
 * @param input - Hull figures, loaded mass, fitted thrusters and ENG pips.
 * @returns The build's {@link MobilityMetrics}, or `null` without thrusters. A mass above
 * the thrusters' maximum returns zero performance rather than a fabricated curve value.
 * @throws {RangeError} If an input is not finite or is outside its documented range,
 * or a thruster curve does not follow the ordering documented by
 * {@link ThrusterCurveParams}.
 * @example
 * ```ts
 * import { mobilityMetrics } from '@elite-dangerous-almanac/core/ships/mobility';
 *
 * mobilityMetrics({
 *   minimumSpeed: 100, maximumSpeed: 220, boost: 320,
 *   minPitch: 34, pitch: 42, minRoll: 110, roll: 110, minYaw: 16, yaw: 16,
 *   mass: 48,
 *   thrusters: {
 *     minMass: 24, optMass: 48, maxMass: 72,
 *     minMultiplier: 0.83, optMultiplier: 1, maxMultiplier: 1.03,
 *   },
 * }).speed; // -> 220
 * ```
 */
export function mobilityMetrics(
    input: MobilityInput & { readonly thrusters: ThrusterParams },
): MobilityMetrics;
export function mobilityMetrics(input: MobilityInput): MobilityMetrics | null;
export function mobilityMetrics(input: MobilityInput): MobilityMetrics | null {
    const pips = input.enginesPips ?? 4;
    requireFiniteRange('mobilityMetrics', 'enginesPips', pips, 0, 4);
    for (const field of [
        'minimumSpeed',
        'maximumSpeed',
        'boost',
        'pitch',
        'roll',
        'yaw',
        'mass',
    ] as const) {
        requireFiniteNonNegative('mobilityMetrics', field, input[field]);
    }
    if (input.minimumSpeed > input.maximumSpeed) {
        throw new RangeError('mobilityMetrics: minimumSpeed must not exceed maximumSpeed');
    }
    for (const [minimum, maximum] of [
        ['minPitch', 'pitch'],
        ['minRoll', 'roll'],
        ['minYaw', 'yaw'],
    ] as const) {
        requireFiniteRange('mobilityMetrics', minimum, input[minimum], 0, input[maximum]);
    }
    if (!input.thrusters) return null;
    validateMassCurve('mobilityMetrics: thrusters', input.thrusters);
    const massCurveMultiplier = curveMultiplier(
        'mobilityMetrics: speed curve',
        input.mass,
        input.thrusters.speedCurve ?? input.thrusters,
    );
    const rotationMassCurveMultiplier = curveMultiplier(
        'mobilityMetrics: rotation curve',
        input.mass,
        input.thrusters.rotationCurve ?? input.thrusters,
    );
    const pipMultiplier = pips / 4;
    const speedAtPips =
        input.minimumSpeed + (input.maximumSpeed - input.minimumSpeed) * pipMultiplier;
    const handlingAtPips = (maximum: number, minimum: number): number =>
        minimum + (maximum - minimum) * pipMultiplier;
    return Object.freeze({
        loadedMass: input.mass,
        speed: speedAtPips * massCurveMultiplier,
        boost: input.boost * massCurveMultiplier,
        pitch: handlingAtPips(input.pitch, input.minPitch) * rotationMassCurveMultiplier,
        roll: handlingAtPips(input.roll, input.minRoll) * rotationMassCurveMultiplier,
        yaw: handlingAtPips(input.yaw, input.minYaw) * rotationMassCurveMultiplier,
        massCurveMultiplier,
        rotationMassCurveMultiplier,
    });
}

File: dotnet/src/EliteDangerousAlmanac/Ships/Mobility.cs
using System;
using EliteDangerousAlmanac.Ships.Internal;

namespace EliteDangerousAlmanac.Ships;

/// <summary>One post-engineering thruster performance curve.</summary>
/// <param name="MinMass">The loaded mass at which performance reaches the maximum multiplier, in tonnes.</param>
/// <param name="OptMass">The loaded mass at which performance is exactly the optimal multiplier, in tonnes.</param>
/// <param name="MaxMass">The largest loaded mass the thrusters can move, in tonnes.</param>
/// <param name="MinMultiplier">The performance multiplier at the maximum mass.</param>
/// <param name="OptMultiplier">The performance multiplier at the optimal mass.</param>
/// <param name="MaxMultiplier">The performance multiplier at the minimum mass.</param>
/// <remarks>
/// Every value must be finite and zero or more, and the masses and the multipliers must follow the
/// order <see cref="IMassCurve"/> documents.
/// </remarks>
public record ThrusterCurveParams(
    double MinMass,
    double OptMass,
    double MaxMass,
    double MinMultiplier,
    double OptMultiplier,
    double MaxMultiplier) : IMassCurve;

/// <summary>A fitted thruster's post-engineering mass curves.</summary>
/// <param name="MinMass">The loaded mass at which performance reaches the maximum multiplier, in tonnes.</param>
/// <param name="OptMass">The loaded mass at which performance is exactly the optimal multiplier, in tonnes.</param>
/// <param name="MaxMass">The largest loaded mass the thrusters can move, in tonnes.</param>
/// <param name="MinMultiplier">The performance multiplier at the maximum mass.</param>
/// <param name="OptMultiplier">The performance multiplier at the optimal mass.</param>
/// <param name="MaxMultiplier">The performance multiplier at the minimum mass.</param>
/// <remarks>
/// Enhanced performance thrusters answer speed and handling on two different curves. A thruster
/// that carries neither <see cref="SpeedCurve"/> nor <see cref="RotationCurve"/> uses its own
/// values for both.
/// </remarks>
public sealed record ThrusterParams(
    double MinMass,
    double OptMass,
    double MaxMass,
    double MinMultiplier,
    double OptMultiplier,
    double MaxMultiplier)
    : ThrusterCurveParams(MinMass, OptMass, MaxMass, MinMultiplier, OptMultiplier, MaxMultiplier)
{
    /// <summary>The speed and boost curve, where it differs from the handling curve.</summary>
    public ThrusterCurveParams? SpeedCurve { get; init; }

    /// <summary>The pitch, roll and yaw curve, where it differs from the speed curve.</summary>
    public ThrusterCurveParams? RotationCurve { get; init; }
}

/// <summary>Everything a mobility calculation needs about one loaded ship.</summary>
/// <param name="MinimumSpeed">The hull's speed at multiplier one and no ENG pips, in metres per second.</param>
/// <param name="MaximumSpeed">The hull's speed at multiplier one and four ENG pips, in metres per second.</param>
/// <param name="Boost">The hull's boost speed at multiplier one, in metres per second.</param>
/// <param name="MinPitch">The hull's pitch rate at no ENG pips, in degrees per second.</param>
/// <param name="Pitch">The hull's pitch rate at four ENG pips, in degrees per second.</param>
/// <param name="MinRoll">The hull's roll rate at no ENG pips, in degrees per second.</param>
/// <param name="Roll">The hull's roll rate at four ENG pips, in degrees per second.</param>
/// <param name="MinYaw">The hull's yaw rate at no ENG pips, in degrees per second.</param>
/// <param name="Yaw">The hull's yaw rate at four ENG pips, in degrees per second.</param>
/// <param name="Mass">The loaded mass, in tonnes: hull, modules, fuel and cargo.</param>
/// <remarks>
/// Every figure must be finite and zero or more. The minimum speed must not be above the maximum
/// speed, and each zero-pip rotation rate must run from zero through its four-pip counterpart.
/// </remarks>
public sealed record MobilityInput(
    double MinimumSpeed,
    double MaximumSpeed,
    double Boost,
    double MinPitch,
    double Pitch,
    double MinRoll,
    double Roll,
    double MinYaw,
    double Yaw,
    double Mass)
{
    /// <summary>The fitted thrusters' post-engineering curve, or <see langword="null"/> for none.</summary>
    public ThrusterParams? Thrusters { get; init; }
}

/// <summary>Speed and rotation rates for one loaded ship, at full ENG.</summary>
/// <param name="LoadedMass">The loaded mass the figures were calculated at, in tonnes.</param>
/// <param name="Speed">The top speed at this mass and four ENG pips, in metres per second.</param>
/// <param name="Boost">The boost speed at this mass, in metres per second.</param>
/// <param name="Pitch">The pitch rate at this mass and four ENG pips, in degrees per second.</param>
/// <param name="Roll">The roll rate at this mass and four ENG pips, in degrees per second.</param>
/// <param name="Yaw">The yaw rate at this mass and four ENG pips, in degrees per second.</param>
/// <param name="MassCurveMultiplier">The speed curve's multiplier at this loaded mass.</param>
/// <param name="RotationMassCurveMultiplier">
/// The rotation curve's multiplier, which differs for enhanced performance thrusters.
/// </param>
/// <remarks>
/// The loaded mass is reported so a caller that never states the mass itself can read what the
/// curve was evaluated at. Against the fitted curve's optimal and maximum masses it is the build's
/// position on that curve.
/// </remarks>
public sealed record MobilityMetrics(
    double LoadedMass,
    double Speed,
    double Boost,
    double Pitch,
    double Roll,
    double Yaw,
    double MassCurveMultiplier,
    double RotationMassCurveMultiplier);

/// <summary>
/// The data-free speed and handling arithmetic for a loaded ship, at full ENG. The figures are the
/// hull's four-pip endpoints with the fitted thruster mass curves applied.
/// </summary>
/// <remarks>
/// <para>
/// The ENG capacitor is a separate story with its own entry point,
/// <see cref="MobilityCapacitor"/>, which interpolates the hull's zero-pip endpoints towards the
/// four-pip ones before the same curves are applied.
/// </para>
/// <para>
/// Ported from EDCD/Coriolis and cross-checked against EDSY's mass-curve calculation. See
/// <c>ATTRIBUTIONS.md</c> for credit and licence terms.
/// </para>
/// </remarks>
public static class Mobility
{
    /// <summary>Resolves a thruster's performance multiplier at a loaded mass.</summary>
    /// <param name="mass">The loaded mass, in tonnes.</param>
    /// <param name="thrusters">The fitted thruster curve, post-engineering.</param>
    /// <returns>
    /// The curve multiplier, and zero above the thrusters' maximum supported mass.
    /// </returns>
    /// <exception cref="ArgumentNullException"><paramref name="thrusters"/> is <see langword="null"/>.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// The mass or a curve value is not finite and zero or more, or the curve masses or
    /// multipliers are not in the order <see cref="IMassCurve"/> documents.
    /// </exception>
    public static double ThrusterMassCurveMultiplier(double mass, ThrusterCurveParams thrusters) =>
        MassCurve.MultiplierAt(mass, thrusters, nameof(mass), nameof(thrusters));

    /// <summary>Calculates a loaded ship's top speed, boost speed and rotation rates at full ENG.</summary>
    /// <param name="input">The hull figures, the loaded mass and the fitted thrusters.</param>
    /// <returns>
    /// The build's metrics, or <see langword="null"/> when no thrusters are fitted. A mass above
    /// the thrusters' maximum answers zero performance rather than a fabricated curve value.
    /// </returns>
    /// <remarks>
    /// <para>
    /// Speed, pitch, roll and yaw are the hull's four-pip endpoints with the fitted thruster's
    /// speed and rotation mass-curve multipliers applied. Boost is independent of the ENG
    /// allocation and uses the speed curve at the loaded mass, so it is the same figure whatever
    /// the pips.
    /// </para>
    /// <para>
    /// A lower allocation is <see cref="MobilityCapacitor.Metrics"/>. It interpolates each figure
    /// from the hull's zero-pip endpoint towards the endpoint used here, and applies the same
    /// curves.
    /// </para>
    /// </remarks>
    /// <exception cref="ArgumentNullException"><paramref name="input"/> is <see langword="null"/>.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// An input is not finite or is outside its documented range, or a thruster curve is not in
    /// the order <see cref="IMassCurve"/> documents.
    /// </exception>
    public static MobilityMetrics? Metrics(MobilityInput input)
    {
        MobilityCurves? curves = MobilityCore.Resolve(input);
        if (curves is null) return null;

        return new MobilityMetrics(
            LoadedMass: input.Mass,
            Speed: input.MaximumSpeed * curves.MassCurveMultiplier,
            Boost: input.Boost * curves.MassCurveMultiplier,
            Pitch: input.Pitch * curves.RotationMassCurveMultiplier,
            Roll: input.Roll * curves.RotationMassCurveMultiplier,
            Yaw: input.Yaw * curves.RotationMassCurveMultiplier,
            MassCurveMultiplier: curves.MassCurveMultiplier,
            RotationMassCurveMultiplier: curves.RotationMassCurveMultiplier);
    }
}
